### PR TITLE
spirv-fuzz: Fuzzer pass that adds access chains

### PR DESCRIPTION
**Pull request by afd**
This change adds a fuzzer pass that sprinkles access chain
instructions into a module at random.  This allows other passes to
have a richer set of pointers available to them, in particular the
passes that add loads and stores.

### CODE REVIEW

File: source/fuzz/fuzzer_pass_add_access_chains.cpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+FuzzerPassAddAccessChains::FuzzerPassAddAccessChains(
+    opt::IRContext* ir_context, FactManager* fact_manager,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, fact_manager, fuzzer_context, transformations) {}
+
+FuzzerPassAddAccessChains::~FuzzerPassAddAccessChains() = default;
+
+void FuzzerPassAddAccessChains::Apply() {
+  MaybeAddTransformationBeforeEachInstruction(
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_add_access_chains.cpp]: Maybe not in this PR, but should this method be renamed? Is it more like ForEachInstructionWithInstructionDescriptor?

File: source/fuzz/fuzzer_pass_add_access_chains.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+        std::vector<opt::Instruction*> relevant_pointer_instructions =
+            FindAvailableInstructions(
+                function, block, inst_it,
+                [](opt::IRContext* context,
+                   opt::Instruction* instruction) -> bool {
+                  if (!instruction->result_id() || !instruction->type_id()) {
+                    return false;
+                  }
+                  switch (instruction->result_id()) {
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_add_access_chains.cpp]: Should this be opcode?

File: source/fuzz/fuzzer_pass_add_access_chains.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                    case SpvOpUndef:
+                      // Do not allow making an access chain from a null or
+                      // undefined pointer.
+                      return false;
+                    default:
+                      break;
+                  }
+                  return context->get_def_use_mgr()
+                             ->GetDef(instruction->type_id())
+                             ->opcode() == SpvOpTypePointer;
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_add_access_chains.cpp]: Could add comment

File: source/fuzz/transformation_access_chain.cpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    return false;
+  }
+
+  // Do not allow making an access chain from a null or undefined pointer, as
+  // we do not want to allow accessing such pointers.  This might be acceptable
+  // in dead blocks, but we conservatively avoid it.
+  switch (pointer->opcode()) {
+    case SpvOpConstantNull:
+    case SpvOpUndef:
+      return false;
```
Comment (paulthomson) [inline source/fuzz/transformation_access_chain.cpp]: Perhaps this is another example where having some kind of "strict assert false" might help in finding bugs, like how I think you accidentally used ->result_id() earlier.

Comment (paulthomson) [reply source/fuzz/transformation_access_chain.cpp]: But I suppose a reducer could simplify pointers...but then we could disable such assertions.

File: test/fuzz/transformation_access_chain_test.cpp [hunk 5]
```diff
@@ -0,0 +1,377 @@
+// Copyright (c) 2019 Google LLC
```
Comment (paulthomson) [inline test/fuzz/transformation_access_chain_test.cpp]: Date?

File: source/fuzz/fuzzer_pass_add_access_chains.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_add_access_chains.cpp]: https://github.com/KhronosGroup/SPIRV-Tools/issues/3184

File: source/fuzz/fuzzer_pass_add_access_chains.cpp (hunk 3 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_add_access_chains.cpp]: Done.

File: source/fuzz/fuzzer_pass_add_access_chains.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_add_access_chains.cpp]: Good catch!

File: source/fuzz/transformation_access_chain.cpp (hunk 4 above)
Comment (afd) [reply source/fuzz/transformation_access_chain.cpp]: For now I think we should assert false in cases where we don't want the reducer or fuzzer to generate certain transformations.  This is one of them, so I have added an assert false.

File: test/fuzz/transformation_access_chain_test.cpp (hunk 5 above)
Comment (afd) [reply test/fuzz/transformation_access_chain_test.cpp]: Fixed.

**Pull request merged**